org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/client/PathWarnings.java

```diff
@@ -0,0 +1,42 @@
+package org.eclipse.lemminx.client;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.lemminx.customservice.XMLLanguageClientAPI;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lsp4j.Command;
+
+public class PathWarnings extends AbstractXMLNotification {
```

Comment:
It might make more sense to name this class `PathWarner`, change `cache` to a `Set`, and store `settingId` as a attribute that is set in the constructor. Then you can make a different `PathWarner` for each setting you want to watch. I think this makes sense in this context, since the only time this class is being used, it is only tracking one setting.

Reply:
> I think it would make more sense to call this class AbstractXMLNotifier, since this generates a notification, but doesn't represent one itself.
> It might make more sense to name this class PathWarner

I like this idea. This means `LimitExceededWarnings` should change to `LimitExceededWarner` too. I will change this soon.

> change cache to a Set, and store settingId as a attribute that is set in the constructor.

I realized that for `LimitExceededWarnings`, the cache is being shared for all features: https://github.com/eclipse/lemminx/blob/c9fa0cc44faa55c80c5d06a5824039be948eec3c/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/client/LimitExceededWarnings.java#L53

Right now, we only have one, so its not a problem. In my latest PR, I separated the cache for each feature using a map (key is the feature, value is a Set), but perhaps its a bit messy. 

If we want one feature = one instance, we must remove this method signature: https://github.com/eclipse/lemminx/blob/c9fa0cc44faa55c80c5d06a5824039be948eec3c/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/client/LimitExceededWarnings.java#L53
and provide the feature in the constructor as you said.

I quickly made a branch that implements that change: https://github.com/xorye/lsp4xml/tree/217catalogs_path_cleaner_cache

Reply:
@xorye I thing your design is very good now. The validation is processed in plugin + AbstractXMLNotification manages cache. 

what is the problem with  public void `onResultLimitExceeded(String uri, LimitFeature feature)` signature?

Reply:
@angelozerr If we follow this suggestion:

> It might make more sense to name this class PathWarner, change cache to a Set, and store settingId as a attribute that is set in the constructor. 

Changing the cache type for `PathWarnings` would change the cache type for `LimitExceededWarnings` too. The cache would only store info for one feature, since the cache would be a `Set<String>` instead of `Map<String, Set<String>>`.

Therefore, it would make `LimitExceededWarnings` work only for one feature, which would change this method signature:
```onResultLimitExceeded(String uri, LimitFeature feature)```
to
```onResultLimitExceeded(String uri)``` 

So, there's no problem with `onResultLimitExceeded(String uri, LimitFeature feature)` right now in this PR, its just that if the suggestion is followed, it does not make sense to provide `feature` in  `onResultLimitExceeded(String uri, LimitFeature feature)`